jina/serve/bff.py

```diff
@@ -0,0 +1,143 @@
+        :param graph_conditions: Dictionary stating which filtering conditions each Executor in the graph requires to receive Documents.
+        :param deployments_disable_reduce: list of Executor disabling the built-in merging mechanism.
+        :param timeout_send: Timeout to be considered when sending requests to Executors
+        :param retries: Number of retries to try to make successfull sendings to Executors
+        :param compression: The compression mechanism used when sending requests from the Head to the WorkerRuntimes. For more details, check https://grpc.github.io/grpc/python/grpc.html#compression.
+        :param runtime_name: Name to be used for monitoring.
+        :param prefetch: How many Requests are processed from the Client at the same time.
+        :param logger: Optional logger that can be used for logging
+        :param metrics_registry: optional metrics registry for prometheus used if we need to expose metrics
+        """
+        topology_graph = self._create_topology_graph(graph_representation, graph_conditions,
+                                                     deployments_disable_reduce, timeout_send, retries)
+        self._connection_pool = self._create_connection_pool(executor_addresses, compression, metrics_registry, logger)
+        request_handler = RequestHandler(metrics_registry, runtime_name)
+
+        self._streamer = RequestStreamer(
+            request_handler=request_handler.handle_request(
+                graph=topology_graph, connection_pool=self._connection_pool
+            ),
+            result_handler=request_handler.handle_result(),
+            prefetch=prefetch,
+            logger=logger,
+        )
+        self._streamer.Call = self._streamer.stream
+
+    def _create_topology_graph(self, graph_description, graph_conditions, deployments_disable_reduce, timeout_send,
+                               retries):
+        # check if it should be in K8s, maybe ConnectionPoolFactory to be created
+        return TopologyGraph(
+            graph_representation=graph_description,
+            graph_conditions=graph_conditions,
+            deployments_disable_reduce=deployments_disable_reduce,
+            timeout_send=timeout_send,
+            retries=retries,
+        )
+
+    def _create_connection_pool(self, deployments_addresses, compression, metrics_registry, logger):
+        # add the connections needed
+        connection_pool = GrpcConnectionPool(
+            logger=logger,
+            compression=compression,
+            metrics_registry=metrics_registry,
+        )
+        for deployment_name, addresses in deployments_addresses.items():
+            for address in addresses:
+                connection_pool.add_connection(
+                    deployment=deployment_name, address=address, head=True
+                )
+
+        return connection_pool
+
+    def stream(self, *args, **kwargs):
```

Comment:
So to use this directly, people users would create Documents out of whatever data their receieve through their own API, and pass those to this, right?

Reply:
this is the one to serve how we already use. I think users should use the `stream_docs` one